packages/nx/src/utils/nx-plugin.ts

```diff
@@ -23,31 +29,78 @@ import {
   findProjectForPath,
 } from '../project-graph/utils/find-project-for-path';
 import { normalizePath } from './path';
-import { join } from 'path';
+import { dirname, join } from 'path';
 import { getNxRequirePaths } from './installation-directory';
 import { readTsConfig } from '../plugins/js/utils/typescript';
+import { NxJsonConfiguration } from '../config/nx-json';
 
 import type * as ts from 'typescript';
 import { retrieveProjectConfigurationsWithoutPluginInference } from '../project-graph/utils/retrieve-workspace-files';
+import { NxPluginV1 } from './nx-plugin.deprecated';
+import { ProjectDependencyBuilder } from '../project-graph/project-graph-builder';
+import { combineGlobPatterns } from './globs';
+
+export type ProjectConfigurationBuilder = (
+  projectConfigurationFile: string,
+  context: {
+    projectsConfigurations: Record<string, ProjectConfiguration>;
+    nxJsonConfiguration: NxJsonConfiguration;
+    workspaceRoot: string;
+  }
+) => {
+  projectNodes?: Record<string, ProjectConfiguration>;
+  externalNodes?: Record<string, ProjectGraphExternalNode>;
+};
+
+export type DependencyLocator = (
+  builder: ProjectDependencyBuilder,
+  context: {
+    /**
+     * The current project graph,
+     */
+    readonly graph: ProjectGraph;
+
+    /**
+     * The configuration of each project in the workspace
+     */
+    projectsConfigurations: ProjectsConfigurations;
+
+    /**
+     * The `nx.json` configuration from the workspace
+     */
+    nxJsonConfiguration: NxJsonConfiguration;
+
+    /**
+     * All files in the workspace
+     */
+    fileMap: ProjectFileMap;
+
+    /**
+     * Files changes since last invocation
+     */
+    filesToProcess: ProjectFileMap;
+  }
+) => void | Promise<void>;
 
-export type ProjectTargetConfigurator = (
-  file: string
-) => Record<string, TargetConfiguration>;
-
-/**
- * A plugin for Nx
- */
-export interface NxPlugin {
+export type NxPluginV2 = {
   name: string;
-  processProjectGraph?: ProjectGraphProcessor;
-  registerProjectTargets?: ProjectTargetConfigurator;
 
   /**
-   * A glob pattern to search for non-standard project files.
-   * @example: ["*.csproj", "pom.xml"]
+   * Provides a map between file patterns and functions that retrieve configuration info from
+   * those files. e.g. { '**\/*.csproj': buildProjectsFromCsProjFile }
    */
-  projectFilePatterns?: string[];
-}
+  processProjectNodes?: Record<string, ProjectConfigurationBuilder>;
```

Comment:
```suggestion
  projectConfigurationsConstructors?: Record<string, ProjectConfigurationsConstructor>;
```

Consider using a tuple instead of a Record